src/sentry/search/snuba/backend.py

```diff
@@ -499,49 +494,53 @@ def snuba_search(start, end, project_ids, environment_ids, tags,
+        # of the normal approximation (Wald) confidence interval formula:
+        #
+        # https://en.wikipedia.org/wiki/Binomial_proportion_confidence_interval
+        #
+        # Effectively if we want the estimate to be within +/- 10% of the real
+        # value with 95% confidence, we would need (1.96^2 * p*(1-p)) / 0.1^2
+        # samples. With a starting assumption of p=0.5 (this requires the most
+        # samples) we would need 96 samples to achieve +/-10% @ 95% confidence.
+
+        limit = options.get('snuba.search.hits-sample-size')
+        offset = 0
+        selected_columns.append(('any', ('rand', ()), 'sample'))
+        sort_field = 'sample'
+        orderby = [sort_field]
+        referrer = 'search_sample'
+    else:
+        # Get the top matching groups by score, i.e. the actual search results
+        # in the order that we want them.
+        orderby = ['-{}'.format(sort_field), 'issue']  # ensure stable sort within the same score
+        referrer = 'search'
+
+    snuba_results = snuba.raw_query(
         start=start,
         end=end,
+        selected_columns=selected_columns,
         groupby=['issue'],
         conditions=conditions,
         having=having,
         filter_keys=filters,
         aggregations=aggregations,
-        orderby=['-' + sort_field, 'issue'],  # ensure stable sort within the same score
-        referrer='search',
-        limit=limit + 1,
+        orderby=orderby,
+        referrer=referrer,
+        limit=limit,
         offset=offset,
+        totals=True,  # needs to have totals_mode=after_having_exclusive so we get groups matching HAVING only
     )
-    metrics.timing('snuba.search.num_result_groups', len(snuba_results.keys()))
-    more_results = len(snuba_results) == limit + 1
-
-    # {group_id -> score,
-    #  ...}
-    group_data = {}
-    for group_id, obj in snuba_results.items():
-        # NOTE: The Snuba utility code is trying to be helpful by collapsing
-        # results with only one aggregate down to the single value. It's a
-        # bit of a hack that we then immediately undo that work here, but
-        # many other callers get value out of that functionality. If we see
-        # this pattern again we should either add an option to opt-out of
-        # the 'help' here or remove it from the Snuba code altogether.
-        if len(required_aggregations) == 1:
-            group_data[group_id] = obj
-        else:
-            group_data[group_id] = obj[sort_field]
+    rows = snuba_results['data']
+    total = snuba_results['totals']['total']
 
-    return (
-        list(
-            sorted(
-                ((gid, score) for gid, score in group_data.items()),
-                key=lambda t: t[1], reverse=True
-            )
-        )[:limit], more_results
-    )
+    metrics.timing('snuba.search.num_result_groups', len(rows))
+
+    return [(row['issue'], row[sort_field]) for row in rows], total
```

Comment:
This return type changed but the docstring still says this returns a boolean in the second spot.